ttnn/cpp/ttnn/operations/data_movement/common/kernels/common.hpp

```diff
@@ -6,9 +6,82 @@
 // It's best to copy and paste the functions in rather than include the header as code size will likely explode
 // Best to separate in to cpp/hpp at some point to avoid the code size explosion but need to figure out the linking
 // issues
+#include <stdio.h>
+#include <cstring>
+#define MASK_64      0xFFFFFFFFFFFFFFC0
+#define OFFSET_64    0x000000000000003F
+#define MASK_16      0xFFFFFFFFFFFFFFF0
+#define OFFSET_16    0x000000000000000F
 
 namespace tt::data_movement::common {
 
+#define max_packet_size 8192
+
+template <uint32_t max_transfer_size, bool only_reads>
+FORCE_INLINE void enhanced_noc_async_read(
+    const uint64_t src_noc_addr, const uint32_t dst_l1_addr, const uint32_t bytes) {
+    // If you do not know the max_transfer_size at compile time write 0 to it.
+    // only reads is true if we ONLY use noc_async_read and all calls to tt_memmove have use_read_datamover as True
+    if constexpr (((max_transfer_size < max_packet_size) && (max_transfer_size != 0)) || only_reads) {
+        noc_async_read_one_packet(src_noc_addr, dst_l1_addr, bytes);
+    } else {
+        noc_async_read(src_noc_addr, dst_l1_addr, bytes);
+    }
+}
+
+template <uint32_t max_transfer_size, bool only_writes>
+FORCE_INLINE void enhanced_noc_async_write(
+    const uint32_t src_l1_addr, const uint64_t dst_noc_addr, const uint32_t bytes) {
+    // If you do not know the max_transfer_size at compile time write 0 to it.
+    // only writes is true if we ONLY use noc_async_read and all calls to tt_memmove have use_read_datamover as False
+    if constexpr (((max_transfer_size < max_packet_size) && (max_transfer_size != 0)) || only_writes) {
+        noc_async_write_one_packet(src_l1_addr, dst_noc_addr, bytes);
+    } else {
+        noc_async_write(src_l1_addr, dst_noc_addr, bytes);
+    }
+}
```

Comment:
Please do not add a new hardcoded define for max_packet_size. We already have `NOC_MAX_BURST_SIZE`.

We also already updated some apis to do this sort of thing, ex noc_async_write is already templated to support using the one packet api if you specify a size less than the max transfer size. You could instead copy how noc_async_write was templated for noc_async_read.

Reply:
It has the max packet size support but not the only_writes support which gives  us just a bit more perf as some of the data mover ops only have reader or writer in a kernel. I think we should keep it as is but will migrate when the only writes optimization is also there. I am changing the max page size to NOC_MAX_BURST_SIZE though.

Reply:
We didn't expose only read/only write flags in the original noc apis because initial testing seemed to show it was mostly functional, but there were some kernels where it seemed to hang even though it should have only been issuing only reads/writes so it was never merged since it risks users running into issues. It's fine if you want to use/expose it outside of dataflow api but you should be careful about exposing a common api for users that hasn't been fully tested. Ex. GS does not support > 8K txns at all, but that is not handled correctly here to ignore the flag for GS.

My suggestion here is to remove you logic regarding max transfer size here since it is already handled in noc_async_write, and instead pass the value down.

Ex
```
template <uint32_t max_transfer_size, bool only_writes>
FORCE_INLINE void enhanced_noc_async_write(
    const uint32_t src_l1_addr, const uint64_t dst_noc_addr, const uint32_t bytes) {
    // If you do not know the max_transfer_size at compile time write 0 to it.
    // only writes is true if we ONLY use noc_async_read and all calls to tt_memmove have use_read_datamover as False
    if constexpr (only_writes) {
        noc_async_write_one_packet(src_l1_addr, dst_noc_addr, bytes);
    } else {
        noc_async_write<max_transfer_size>(src_l1_addr, dst_noc_addr, bytes);
    }
}
```